infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -725,7 +742,23 @@ public void append(LogAddress logAddress, LogData entry) {
             SegmentHandle fh = getSegmentHandleForAddress(logAddress);
             if (fh.getKnownAddresses().containsKey(logAddress.address) ||
                     fh.getTrimmedAddresses().contains(logAddress.address)) {
-                throw new OverwriteException();
+                if (entry.getRank()==null) {
+                    throw new OverwriteException();
+                } else {
+                    try (MultiReadWriteLock.AutoCloseableLock ignored = segmentLocks.acquireWriteLock(fh.getSegment())) {
+                        Optional<Boolean> canAppend = isAppendPermitted(logAddress, entry);
+                        if (canAppend.isPresent()) {
+                            if (canAppend.get()) {
+                                AddressMetaData addressMetaData = writeRecord(fh, logAddress.address, entry);
+                                fh.getKnownAddresses().put(logAddress.address, addressMetaData);
+                            } else {
+                                throw new OverwriteException();
+                            }
+                        } else {
+                            log.debug("Duplicate requests for {} ", logAddress);
```

Comment:
Do you care about the case where you have the same ranks, but the data in the entry differs ? would checking that be a good way to find out if some bad happened?